In [50]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [74]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv("Resources/lending_data.csv")
# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [75]:
#Finding out how many rows are duplicated data
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
77531    False
77532     True
77533     True
77534    False
77535     True
Length: 77536, dtype: bool

In [53]:
#dropping the duplicated rows
df = df.drop_duplicates()

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [56]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = df["loan_status"]


# Separate the X variable, the features
X = df.drop('loan_status', axis=1)

In [57]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [58]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [59]:
# Check the balance of our target values
y.value_counts()

0    3653
1    1576
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [60]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [61]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
clf = LogisticRegression(random_state=1).fit(X_train, y_train)


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [62]:
# Make a prediction using the testing data
y_pred = clf.predict(X_test)

In [63]:
y_test.value_counts()

0    909
1    399
Name: loan_status, dtype: int64

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [64]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test, y_pred)

0.8937359901403674

In [65]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_pred)

array([[832,  77],
       [ 51, 348]])

In [66]:
# Print the classification report for the model
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.92      0.93       909
           1       0.82      0.87      0.84       399

    accuracy                           0.90      1308
   macro avg       0.88      0.89      0.89      1308
weighted avg       0.90      0.90      0.90      1308



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Although highly accurate, the model predicts low-risk successfully, but high risk prediction was subpar. The data set was not well balanced and over sampling is needed.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [67]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=42)

# Fit the original training data to the random_oversampler model
X_res, y_res = ros.fit_resample(X_train, y_train)

In [68]:
# Count the distinct values of the resampled labels data
y_res.value_counts()

0    2744
1    2744
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [70]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
clf_res = LogisticRegression(random_state=1).fit(X_res, y_res)

# Make a prediction using the testing data
y_pred = clf_res.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [71]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, y_pred)

0.943464271239154

In [72]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_pred)

array([[829,  80],
       [ 10, 389]])

In [73]:
# Print the classification report for the model
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.91      0.95       909
           1       0.83      0.97      0.90       399

    accuracy                           0.93      1308
   macro avg       0.91      0.94      0.92      1308
weighted avg       0.94      0.93      0.93      1308



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Using the oversampled data the logistic regression model correctly predicts 829 healthy and 389 high-risk loans.  This shifts the balance in the possibilty of falsely classifying a low-risk loan as a high risk loan considerably while maintaining 94% accuracy. 